### PR TITLE
How we tell our stories design

### LINKED ISSUE #61: Implement HWTOS design as a variation on the playlist label

### PR DESCRIPTION
**Pull request by cogat**
Resolves #61

I've added an alternative template for How We Tell Our Stories -type labels, and refactored the javascript to support this, and to allow key controls for debugging.

### Relevant design files
Layout

### Testing with local XOS
1. Set up a playlist for playlist label containing max 7 items, with videos and labels. You may want to make some works indigenous and have a title_annotation. You way want to put markdown in some of the description fields. Note the ID and an auth token.
2. edit `config.env` to include:
```
XOS_API_ENDPOINT=http://[YOUR_IP]:8000/api/
AUTH_TOKEN=[YOUR AUTH TOKEN]
XOS_MEDIA_PLAYER_ID=1
XOS_PLAYLIST_ID=[PLAYLIST_ID]
LABEL_TEMPLATE=up_next.html
```
3. Run `cd development && docker-compose up` to run the playlist app
4. View the label at localhost:8081 - ensure your labels are shown and design looks like the attached. Ensure the durations of your videos are properly reflected.
5. Press 0-9 to move the progress bar to 0-90% and see the times update. Press 't' to simulate a tap.
6. Press right arrow to move to next item. Notice that the browser # anchor updates with label ID. Refresh the browser to see that the current label is reloaded (albeit with flash of initial label - removing that can be for another ticket).
7. Remove `LABEL_TEMPLATE` env var, and restart the docker-compose command, and see that the standard label renders.

### Testing with playlist hardware (needs VPN connection to XOS)
1. Set up a playlist in XOS; note the ID.
2. Deploy a playlist label to use the `s__playlist-label-pi-4` app. Configure the device with `XOS_PLAYLIST_ID`, and `LABEL_TEMPLATE` as above.
3. Observe that the playlist works as expected.
4. If a keyboard is connected, the keys work as above.
5. If a media player is connected, the playlist updates to match
6. If a lens reader is connected, the lens reader captures the currently-shown label and the 'collect' animation displays properly.

### CODE REVIEW

File: app/static/playlist.js [hunk 1]
```diff
@@ -105,54 +142,107 @@ export default class PlaylistLabelRenderer {
...
+
+      const id = "#up_next_label_" + i;
+      try {
+        document.querySelector(id + " .time_to_wait").innerHTML = num_minutes + unit;
+      } catch(err) {}
+      time_to_wait += items[i].video.duration_secs;
+    }
+  }
+
+  autoUpdateProgress(slf) {
```
Comment (sighmon) [inline app/static/playlist.js:239]: I don't mind this for dev mode, but I think it should have to be explicitly turned on. My progress bar jumps backwards in time for some videos, so it's not great.

File: app/static/playlist.js [hunk 2]
```diff
@@ -105,54 +142,107 @@ export default class PlaylistLabelRenderer {
...
+        document.querySelector(id + " .title").innerHTML = i + '. ' + label.title;
+        document.querySelector(id + " .subtitles").innerHTML = label.subtitles;
+      } catch(err) {}
+    }
+  }
+
+  updateProgress() {
+    const progressBar = document.getElementById("progress-bar");
+    const playbackPosition = this.state.playbackPosition;
+    progressBar.style.width = `${playbackPosition * 100}%`;
```
Comment (sighmon) [inline app/static/playlist.js:218]: Do you want to pull Harry's progress bar code in here to make it smoother? https://github.com/ACMILabs/playlist-label/pull/17/files#diff-9c4ac97c58888f532e32895a3c9de5f6R156

File: app/static/playlist.js (hunk 1 above)
Comment (cogat) [reply app/static/playlist.js:250]: I have turned it off, unless the 'ignore_media_player' flag is set. I haven't yet implemented a clean way to set it - it is always off in this repo, and always on in XOS preview.

File: app/static/playlist.js (hunk 2 above)
Comment (cogat) [reply app/static/playlist.js:229]: There's a different ticket for that; I'll leave it for now.

**Comment by sighmon** [comment]
Note: I've now got this code running on [MA-08-LR01-RP01](https://dashboard.balena-cloud.com/devices/c6f9d5d322a7e3a8b04ce971a62b6869d83e7868743a951ef422377b76c4f7/summary) in [s__playlist-label-pi-4](https://dashboard.balena-cloud.com/apps/1529802/devices)

**Review by sighmon — COMMENTED** [review]
Fixed a test and updated a couple of little things, but good to go now.

**Pull request merged**